Functions as first-class types.

chriseth opened the pull request:
This adds function types to the types available to the user. There are internal and external functions, the former of which cannot be used in external function calls (one of the reasons is that it would not be possible to evaluate them).

Fixes https://github.com/ethereum/solidity/issues/910, #1353.

Still to do:
- [x] documentation
- [x] discussion if this is the syntax we want
- [x] discussion if `internal` should be the default for function types
- [x] discussion if that is the encoding we want (left aligned `<function id><address>`)
- [x] special case in parser for function types for storage variables (distinguish from fallback function)
- [x] storing function types in storage (and tests for that)
- [x] arrays of functions (also in external function calls)
- [x] squash the commits
- [x] check that delete works on function types (in storage and on stack)
- [x] internal functions should survive the transition from constructor to runtime context in storage.
- [x] zero value for storage function should throw (for internal function, this is a problem because zero is actually a valid jumpdest) solution: invert the value (better than e.g. adding 1 because it can be equally applied to construction and runtime value)
- [x] modify the grammar file
- [x] use new proper ABI type for functions

chriseth committed:
Function types.

chriseth committed:
Code generator for function types.

chriseth commented:
This PR already provides basic functionality, you can pass along internal functions in internal function calls and evaluate them and you can pass along external functions in external function calls.

chriseth commented:
One suggestion concerning notation: TypeScript would use `(uint, uint) => (uint)` for a function type. It would not be too obvious where to place the `external` / `internal` keyword, but it would be much shorter.

VoR0220 commented:
agreed with the above suggestion as it also makes it more readable and able to discern which function is the "root" function and which are the parameters or in scope functions.

axic commented:
>  `function eval(function(uint) external returns (uint) x, uint a) returns (uint) {`

I think we should come up with a better way of representing this, maybe requiring parentheses around could work: `(function(uint) external returns (uint)) x`

chriseth commented:
Parentheses might improve readability but it is not too logical (especially consider the two cases where you have only the type and where you also have a variable name), so I would not force the requirement.

axic commented:
Actually this is the time where the `alias` / `typedef` keyword could be useful :smiley:

VoR0220 commented:
I think `function eval(function(uint) external => (uint) x, uint a) returns (uint) {` would be better....the arrow makes it much more readable...perhaps some enforcement by the compiler that function types go at the end of the args? Or at the very least put that in the style guide...

pirapira commented:
@VoR0220 ,  @chriseth and I were discussing if the keywords `external` `internal` `pure` or `view` should come before/after the arrow `=>`.  Things before the arrow should talk about the input, and things after the arrow should talk about the output, but `external` talks about neither.  Maybe `external` and `internal` should come before `function`, so:

```
function eval(external function(uint) => (uint) x, uint a) returns (uint) {
              ^^^^^^^^
```

VoR0220 commented:
yeah but then you have to think about how that syncs with the rest of language...hmmmm...

axic commented:
> Parentheses might improve readability but it is not too logical (especially consider the two cases where you have only the type and where you also have a variable name)

`(function())` and `(function()) x`

I don't think it looks that horrible. (Could be better.) The aim is to have the same syntax as current function definitions.

> so I would not force the requirement.

Even if it is not enforced, but is accepted by the parser, it could go into the style guide.

axic commented:
>  Maybe `external` and `internal` should come before function

I like this as a future change, which applies to actual function definitions too. However, where do you put `payable`, `view`, `pure` and the modifiers?

pirapira commented:
@axic `payable` `view` and `pure` can be in the same position as `external` `internal` i.e. before `function`.  I doubt we should include general modifiers in the type of functions.

axic commented:
> I doubt we should include general modifiers in the type of functions.

I'm thinking about using the same notation (whatever we come up with) for both declarations and variables.

pirapira commented:
@axic When you receive a function `g` in `function f (function g() onlyOwner)` how do we check whether `g` behaves as `onlyOwner`?  Do we always have access to the Solidity source?

axic commented:
@pirapira sorry, the quote was misleading. I don't think modifiers apply to function types. However I would like to use the same notation in general (with the exception that modifiers, `public` and `private` are not valid for types).

pirapira commented:
@axic I agree that we should make the notations uniform here and in function declarations.

chriseth committed:
Function type state variables.

chriseth committed:
Fix some type checks and tests for internal / external function parameters.

chriseth committed:
External functions in storage.

pirapira committed:
Add tests around calling functions returning functions returning functions

pirapira committed:
Add a parser test for arrays containing functions

pirapira committed:
Add a test around storing functions in an array

pirapira commented on libsolidity/ast/ASTJsonConverter.cpp:
```diff
@@ -225,6 +225,15 @@ bool ASTJsonConverter::visit(UserDefinedTypeName const& _node)
+{
+	addJsonNode(_node, "FunctionTypeName", {
+		make_pair("payable", _node.isPayable()),
+		make_pair("constant", _node.isDeclaredConst())
```
Maybe visibility can be added here.

pirapira commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -657,6 +683,14 @@ void CompilerUtils::convertType(Type const& _typeOnStack, Type const& _targetTyp
+	{
+		if (funType->location() == FunctionType::Location::Internal)
+		{
+			m_context << m_context.errorTag();
```
Why is this a runtime error, not an error reported at the compilation time?

pirapira replied:
Ah, now I understand, you need to initialize array elements that contain internal functions, with some values.

chriseth replied:
Yes, this is also used when you do `function() x; x();`

pirapira replied:
In my taste I would disallow `function() x` without an initialization.  Also, in my taste I would disallow array extension without giving elements, but I realized that's a bigger change and for now we allow array extension without giving elements.

pirapira commented on libsolidity/ast/Types.cpp:
```diff
@@ -1961,6 +2008,26 @@ MemberList::MemberMap FunctionType::nativeMembers(ContractDefinition const*) con
+	if (m_location != Location::External && m_location != Location::Internal)
+		return TypePointer();
+	if (m_location != Location::External)
+		return TypePointer();
```
The two checks above look redundant (the first one catches a subset of what the second one catches).

chriseth replied:
Should be fixed now.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
+	BOOST_CHECK(callContractFunction("f()") == encodeArgs(u256(2)));
+}
+
+BOOST_AUTO_TEST_CASE(array_of_functions)
```
It seems to be a mapping, not an array

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
+	BOOST_CHECK(callContractFunction("checkSuccess()") == encodeArgs(true));
+}
+
 // TODO: arrays, libraries with external functions
```
We should also test that `var x = new function()[](20); x[2]();` correctly throws.

chriseth replied:
Oh and do we support comparison between functions? I think we should.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
+{
+	char const* sourceCode = R"(
+		contract Flow {
+			bool success;
```
Use `bool public success`, then you do not need `checkSuccess`

pirapira replied:
I was lazy to look up the name of the accessor function.

chriseth replied:
it's just `success()`

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
+			}
+
+			function f () {
+				if (0 == steps[msg.sender])
```
`steps` seems to be a typo.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
+
+	compileAndRun(sourceCode, 0, "C");
+	BOOST_CHECK(callContractFunction("checkSuccess()") == encodeArgs(false));
+	callContractFunction("f()");
```
Please check the return value.

pirapira replied:
I was lazy to lookup what I should expect as the return value from a function that does not return a value.

chriseth replied:
`BOOST_CHECK(callContractFunction("f()") == encodeArgs())`

chriseth commented on test/libsolidity/SolidityParser.cpp:
```diff
@@ -1311,6 +1311,31 @@ BOOST_AUTO_TEST_CASE(function_type_as_parameter)
+	BOOST_CHECK(successParse(text));
+}
+
+BOOST_AUTO_TEST_CASE(array_of_functions)
```
Mapping

chriseth committed:
Changelog entry and small fixes.

chriseth committed:
Part of the documentation.

chriseth committed:
Change alignment.

chriseth committed:
Tests.

chriseth committed:
Documentation examples.

chriseth commented:
In the meantime, I assume that we at least settled for the basic notation for function types with the option to also introduce a short notation in the future. I also think that the default value of "internal" for function types is fine.

During testing, I found an unfortunate situation I would like to get opinions on:

```
contract c {
  function (uint) internal returns (uint) x = a;
  function a(uint) internal returns (uint) { return 2; }
}
```

The assignment `x=a` will be done at construction time, but the value of the identifier `a` is different at construction time than it is at runtime (because the code is different). I don't see a real language-based solution that this problem, so we should find a way to access internal functions at construction time that are also valid at runtime. We know which of the internal functions are accessed, so it might make sense to create a "trampoline" at a fixed point which has a special flag so it will not be moved by the optimizer.

pirapira commented:
@chriseth you can use `CODESIZE - 100` to access the same function at the construction time and at runtime in a uniform way.

chriseth commented:
@pirapira proposed to represent internal functions as a pair of `(construction_time_label, runtime_label)`. Both can be limited to 4 bytes. If a function is accessed at runtime, it is fine to leave hte first part as zero. If it is accessed at construction time, both are filled. If a function is called at runtime, the upper bits are zeroed out, if it is called at construction time, the lower bits are zeroed out.

chriseth committed:
Bugfix in code generator.

pirapira committed:
test: add a test for storing an internal function in the constructor and then using the stored function in runtime

pirapira commented on docs/types.rst:
```diff
@@ -266,6 +266,118 @@ to and from all integer types but implicit conversion is not allowed.
+Function Types
+--------------
+
+Functions can be assigned to variables and passed on together with function calls.
```
`passed on together with` -> `passed as arguments to`

pirapira replied:
I want another sentence here.  `The variables and the function arguments that contain functions have function types.  A function type specifies the argument types and the return type of the function.  Moreover (continues next line)`

chriseth replied:
Agreed that this needs more explanation. I modified your sentence slightly.

pirapira commented on docs/types.rst:
```diff
@@ -266,6 +266,118 @@ to and from all integer types but implicit conversion is not allowed.
+Internal functions can only be used inside the current contract (more specifically,
+inside the current code unit, which also includes internal library functions)
+because they cannot be executed outside of the
+context of the current function. Calling an internal function is realized
```
`outside of the context of the current function` -> `outside of the context of the current contract`

pirapira commented on docs/types.rst:
```diff
@@ -266,6 +266,118 @@ to and from all integer types but implicit conversion is not allowed.
+
+Function types are notated as follows:
+
+    function (<argument types>) internal / external returns (<return types>)
```
Maybe `{internal|external}` to be explicit about what can be altered.

pirapira commented on docs/types.rst:
```diff
@@ -266,6 +266,118 @@ to and from all integer types but implicit conversion is not allowed.
+
+    function (<argument types>) internal / external returns (<return types>)
+
+As always, the ``returns (<return types>)`` is optional.
```
What is the interpretation when `returns` is missing?

pirapira commented on docs/types.rst:
```diff
@@ -266,6 +266,118 @@ to and from all integer types but implicit conversion is not allowed.
+      )
+        returns (uint r)
+      {
+        r = self[0];
```
Maybe check the length of `self` before accessing an element.

chriseth replied:
It will throw an exception if the array is too short.

chriseth committed:
Tests for uninitialized storage functions.

chriseth committed:
Implement uninitialized storage functions.

chriseth committed:
Updates to the documentation.

chriseth committed:
Added function types to the grammar.

pirapira commented:
Currently the build is broken for me (at commit 59e8c63f183e0f62e9b45049c1b0157d4f5f2ce8).

pirapira committed:
codegen: add a compilation mode and a runtime context to CompilerContext

chriseth commented on libsolidity/codegen/Compiler.h:
```diff
@@ -35,7 +35,9 @@ class Compiler
 		m_optimize(_optimize),
-		m_optimizeRuns(_runs)
+		m_optimizeRuns(_runs),
+		m_context(CompilationMode::Creation, &m_runtimeContext),
```
Not sure it is a good idea to take a reference of a not yet initialized variable.

pirapira replied:
I'm assuming that the object is not seen before the constructor finishes.

chriseth commented on libsolidity/codegen/CompilerContext.h:
```diff
@@ -37,13 +37,24 @@ namespace dev {
 class CompilerContext
 {
 public:
+	CompilerContext(CompilationMode _mode, CompilerContext* _runtimeContext = nullptr) :
```
A less explicit way of doing the same thing would be to use the null-ness of `_runtimeContext` as the switch.

pirapira replied:
I chose to have redundancy so that it's more difficult to misuse the class.

chriseth commented on libsolidity/codegen/ContractCompiler.h:
```diff
@@ -38,11 +38,11 @@ namespace solidity {
 		m_optimise(_optimise),
 		m_context(_context)
 	{
-		m_context = CompilerContext();
```
I know this is not part of this change, but what is the purpose of `_context` if it is overwritten in `m_context`?

pirapira replied:
This assignment has a side-effect similar to `_context = CompilerContext()`.  The effect is visible from outside of this class.

axic commented on libevmasm/Assembly.cpp:
```diff
@@ -485,6 +485,7 @@ LinkerObject const& Assembly::assemble() const
 			break;
 		case Tag:
 			tagPos[(unsigned)i.data()] = ret.bytecode.size();
+			assertThrow(ret.bytecode.size() < 0xffffffffL, AssemblyException, "Tag too large.");
```
Can we pull this into a separate PR and merge it now?

chriseth replied:
I don't see it as so critical.

axic replied:
You don't see 16 Mb bytecodes flying around? ⛑

axic commented on libsolidity/codegen/ContractCompiler.cpp:
```diff
@@ -293,6 +293,7 @@ void ContractCompiler::appendCalldataUnpacker(TypePointers const& _typeParameter
 	{
 		// stack: v1 v2 ... v(k-1) base_offset current_offset
 		TypePointer type = parameterType->decodingType();
+		solAssert(type, "No decoding type found.");
```
Can we also pull this with other assert?

chriseth replied:
I'm planning to merge this today or at latest tomorrow.

axic replied:
Are all the tickboxes done? (See on the top.)

axic commented:
Wondering if `payable` should be part of the signature or not.

chriseth committed:
delete for function types

chriseth committed:
Fix tests.

chriseth commented:
@axic yes, I think payable already is part of the signature, I just forgot to mention it in the documentation and the grammar.

axic commented:
>  discussion if that is the encoding we want (left aligned `<function id><address>`)

Will we also add conversion from `function` type to `address` since that seems natural? A recipient of a function type could decide not to call it if it refers to a third party address for example.

axic commented:
One shortcoming I see is a limitation we have in the ABI in general: return types are not part of the signature. I would like that to be addressed in the future.

axic commented:
> By default, function types are internal, so the ``internal`` keyword can be omitted.

In the oracle example, what happens if an ``internal`` function is passed around?

axic commented on docs/types.rst:
```diff
@@ -266,6 +266,122 @@ to and from all integer types but implicit conversion is not allowed.
+function type should not return anything, the whole ``returns (<return types>)``
+part has to be omitted.
+
+By default, function types are internal, so the ``internal`` keyword can be
```
In the oracle example, what happens if an `internal` (or unmarked) function is passed around?

chriseth replied:
You will get a type error

axic replied:
Even if the `external` is removed from the Oracle example? So all the parameters default to `internal`.

chriseth replied:
Ah now I see what you mean. The default for a function type is "internal". Furthermore, the default for a function implementation is also "internal", unless it is defined at contract level scope, which makes it external and internal at the same time, i.e. public.

The type of a contract level function f is internal if you access it only as `f` and it is external if you access it as `this.f` - I will update the documentation to mention that.

chriseth commented:
Return types are not part of the abi signature: Let's discuss that somewhere else.

chriseth committed:
Mention "payable" in the documentation.

chriseth committed:
Disallow payable internal functions.

chriseth committed:
Test passing functions as arrays to other contracts.

chriseth commented:
@pirapira discovered that 0 is actually not a valid jump destination, it was like that way before frontier.

chriseth committed:
ABI: Use external function.

chriseth committed:
Fix tests.

chriseth committed:
Stored combined creation and runtime tags.

Includes a change to Assembly to allow tags from sub-assemblies to be
used.

Sorry, this get a bit bigger than I thought.

axic commented:
> >    discussion if that is the encoding we want (left aligned <function id><address>)
> 
> Will we also add conversion from function type to address since that seems natural? A recipient of a function type could decide not to call it if it refers to a third party address for example.

This reminds me: it should be right aligned (left padded) to support conversion by masking (and not shifting).

That would also change its output type from `bytes24` to `uint192`.

chriseth commented:
@axic the main reason I changed it to `bytes24` was that it is much closer to a `bytes24` than it is to `uint192`. Furthermore, conversion by masking does not help much, you always have to shift, too.

axic commented:
As an example, if it is right aligned: `Test(0xca35b7d915458ef540ade6068dfe2f44e8fa733c).confirm(bytes32)` becomes: `00000000797af627ca35b7d915458ef540ade6068dfe2f44e8fa733c`.

Typecasting that to `address` (`uint160`) should gives us the address.

chriseth commented:
Simple typecasting to address is exactly one danger I would like to avoid. How the higher order bits are handled for address is unspecified, so if you pass an external function to a function that assumes an address, it will perform the conversion without warning and just ignore the function signature part.

chriseth committed:
Converted sub assembly to smart pointer.

pirapira committed:
test: add a test about external function type taking/returning internal functions

chriseth committed:
Check that no internals are used in any external function type.

axic commented:
What is unspecified about higher order bits for addresses? Also the conversion from function to address should not be implicit.

chriseth committed:
Simple peephole optimizer that is activated even if not requested.

chriseth committed:
Fix parser for function type disambiguity.

chriseth committed:
Fix tests.

chriseth committed:
fixup! Simple peephole optimizer that is activated even if not requested.

chriseth commented:
@axic I'm talking about using an external function type in the call data to a contract that expects address.

chriseth committed:
Fix setting the tag.

chriseth commented:
Proposal by @axic: encode functions as left-aligned `<address><function id>` because the address is the more higher-level construct.

chriseth committed:
Disable broken tests that are not useful.

chriseth commented:
Discussion about the ABI resulted in:
- we will use `<address><function id>` (left aligned) as encoding type, potentially using the zero bytes after that as runtime type information at a later point in time
- the ABI type will be `function` (new type name). We do not encode the parameter types because the ABI is only used to disambiguate overloaded functions and specify the encoding. Encoding the full type information of the function is too detailed - we also do not do that for a lot of other things (Contract, for example).

chriseth committed:
Change encoding to address-funid and add "function" as ABI type.

chriseth committed:
Fix documentation formatting.

chriseth commented:
@axic please comment on the latest changes.

axic commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -590,9 +591,19 @@ void CompilerStack::compileContract(
+		// In some cases (if the constructor requests a runtime function), it is not
+		// possible to compile the clone.
+
+		// TODO: Report error / warning
```
Why are we hiding the error here?

chriseth replied:
There is no easy way to report it (at this point, only throwing an exception would work, but this is exactly what we want to avoid) and clones are an undocumented feature anyway.

axic commented:
We should have a test in `SolidityABIJSON` (and maybe in `ASTJSON` too).

axic commented on docs/types.rst:
```diff
@@ -266,6 +266,127 @@ to and from all integer types but implicit conversion is not allowed.
+on it.
+
+If external function types are used outside of the context of Solidity,
+they are converted into the ``bytes24`` type.
```
Shouldn't this be `function` now instead of `bytes24`?

chriseth committed:
JSON tests.

chriseth committed:
Some more tests.

chriseth committed:
Update documentation.

chriseth commented:
Updated documentation and added test.

pirapira commented on docs/types.rst:
```diff
@@ -266,6 +266,128 @@ to and from all integer types but implicit conversion is not allowed.
+Function Types
+--------------
+
+Function types can be used to assign functions to variables and passing them
```
`passing them to` and `return them from` do not balance.  I think it's either `pass them to and return them from` or `passing them to and returning them from`.

chriseth committed:
Documentation: Style update

chriseth commented:
@axic @pirapira ready to merge?

pirapira commented on libsolidity/ast/Types.cpp:
```diff
@@ -1699,7 +1715,7 @@ TypePointer TupleType::closestTemporaryType(TypePointer const& _targetType) cons
 	m_location(_isInternal ? Location::Internal : Location::External),
 	m_isConstant(_function.isDeclaredConst()),
-	m_isPayable(_function.isPayable()),
+	m_isPayable(_isInternal ? false : _function.isPayable()),
```
Can we invoke an assertion failure if a function is internal and payable at the same time?

chriseth replied:
I will try.

chriseth replied:
This is the constructor that creates a function type out of a function definition. If the function is `public`, it can be payable, and the resulting internal function type will not be payable.

pirapira commented on libsolidity/ast/Types.cpp:
```diff
@@ -1800,6 +1816,33 @@ FunctionType::FunctionType(EventDefinition const& _event):
+	m_isPayable(_typeName.isPayable())
+{
+	if (_typeName.isPayable())
+		solAssert(m_location == Location::External, "Internal payable function type used.");
```
Can we also assert that one cannot have `isPayable` and `isConstant` at the same time?

chriseth replied:
That's why I really would like to combine those flags into a single enum or at least a single class that has internal checks for these restrictions.

pirapira commented on libsolidity/ast/Types.cpp:
```diff
@@ -1800,6 +1816,33 @@ FunctionType::FunctionType(EventDefinition const& _event):
+	{
+		solAssert(t->annotation().type, "Type not set for return parameter.");
+		solAssert(
+			m_location != Location::External || t->annotation().type->canBeUsedExternally(false),
```
The assertion reads strange to me.  Maybe `m_location == External /* then it's ok */ || ...`

chriseth replied:
This is a logical implication, but I will rewrite it using `if`.

pirapira commented on libsolidity/ast/Types.cpp:
```diff
@@ -1875,22 +1918,65 @@ bool FunctionType::operator==(Type const& _other) const
+		name += "payable ";
+	if (m_location == Location::External)
+		name += "external ";
+	name += "returns (";
```
The `returns` clause should not exist when the number of return values is zero.  (So I read somewhere.)

axic commented on libevmasm/Assembly.cpp:
```diff
@@ -75,17 +79,17 @@ string Assembly::out() const
+unsigned Assembly::bytesRequired(unsigned subTagSize) const
 {
-	for (unsigned br = 1;; ++br)
+	for (unsigned tagSize = subTagSize;; ++tagSize)
```
It looks weird that there's no spacing between the two semicolons.

axic commented on libsolidity/codegen/CompilerUtils.cpp:
```diff
@@ -133,6 +133,16 @@ void CompilerUtils::storeInMemoryDynamic(Type const& _type, bool _padToWordBound
+		dynamic_cast<FunctionType const&>(_type).location() == FunctionType::Location::External
+	)
+	{
+		solAssert(_padToWordBoundaries, "Non-padded store for function not implemented.");
```
Can this be moved to the `UnimplementedFeature` exception?

chriseth committed:
Some more assertions and style changes.

chriseth commented:
Updated.

pirapira reviewed (approved):
I think this one is ready to merge.
